### ISSUE TITLE
Specify import name space merging (with and without a prefix)

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#380]
It seems that section 'Imports' does _not_ specify that multiple declarations with the same name can be imported (by using the same prefix for several imports, or by using no prefix with several libraries), and name clashes are accepted, such that an error only arises when the name is used.

The spec must describe that behavior (currently I think it just implicitly makes that kind of name clash an error, based on the normal treatment of clashes in name spaces).

Also, it is suggested [here](https://github.com/dart-lang/language/blob/54c7e33580df7d43b98ad30022c630a7c5b2acf8/specification/dartLangSpec.tex#L13976) that $N$ can be added to the exported namespace more than once. Clarify this to ensure that namespaces always follow the rule that a name has at most one binding.

**Comment by eernstg** [comment]
@lrhn, @leafpetersen, @munificent, do you agree that the tools should continue to have this behavior, and the spec just needs to specify it?

**Comment by leafpetersen** [comment]
I think there are two slightly different things that as best I can tell aren't specified:
  - First, that if I import two different libraries with the same prefix, the resulting prefix can be used to access different symbols from *both* libraries
  - Second, if I import two different libraries with or without a prefix, and both libraries define the same symbol, there's no error unless that symbol is referenced (either directly, or via an export of the libraries in question)

Both of these seem like reasonable behavior to me.

**Comment by munificent** [comment]
I agree with Leaf. In particular:

> First, that if I import two different libraries with the same prefix, the resulting prefix can be used to access different symbols from both libraries

We have a large amount of internal code today that relies on this behavior, so *not* allowing this would be a real breaking change. I find the pattern a little odd, but it seems reasonable for the way it's used. Basically, they import all generated protobuf libraries into the same `pb` prefix.

> Second, if I import two different libraries with or without a prefix, and both libraries define the same symbol, there's no error unless that symbol is referenced (either directly, or via an export of the libraries in question)

Yes. I recall discussion around this exact point early in Dart's history and I believe the consensus then was that this was the desired behavior. I think it was originally an error but it proved too frustrating when you imported two libraries that both exported the same declaration from a third library and it yelled at you even though you were getting the same declaration twice.

So I think the intent has been for a very long time that this is safe and allowed.

My understanding is that to have an import collision, you need all of:

* To import two *different* declarations under the same name.
* To actually refer to that declaration in your code.

In other words, it's an error to *use an identifier that can't be unambiguously resolved to a single declaration.*

**Comment by eernstg** [comment]
Very good, your comments match up precisely with the intentions behind my description. I think the "late name clash" can be specified by saying that the introduction of any non-first binding for a given name `n` _replaces_ the binding for `n` in the given namespace by a new binding to a special `NAME_CLASH` value, and any lookup which finds that value is an error. Given that this is only used for namespaces where all lookups are resolved statically, we know that this is a compile-time error.

**Comment by munificent** [comment]
>  by a new binding to a special NAME_CLASH value, and any lookup which finds that value is an error.

Very clever. I really like this. :)

**Comment by lrhn** [comment]
I'm slightly worried by the imperative sound of "replacing". I'd prefer if we can keep the definitions of import scopes purely declarative. We can still use NAME_CLASH as a marker

Something like:

> An import scope is a set of either declarations, a special NAME_CLASH marker, or a special `loadLibrary` function introduced for deferred imports.
>
> Each import in a library produces declarations for an import scope.
> 1. An import with no `as prefix` marker introduces declarations for the *default* import scope.
> 1. An import with an `as prefix` marker introduces declarations for the named import scope denoted by `prefix` for that library.
>
> (A deferred import must use a prefix name not shared with any other import).
>
> For each import scope *s* available to a library (the default scope and each named import scope mentioned by an import in the library), let *L<sub>s</sub>* be the set of imports introducing declarations into that scope.
> For each import *l* in *L<sub>s</sub>*, let *D<sub>l</sub>*, the declarations introduced by *l*, be the set of declarations in the export scope of the library imported by *l* which are not hidden by a `show` or `hide` modifier on the import.
> If *L<sub>s<sub>* contains a deferred library, then it is necessarily a singleton set (deferred import scopes cannot be shared between imports). Then *s* contains a special method with signature `Future<void> loadLibrary()`.
> Let *N<sub>l<sub>* be the names of the declarations of *D<sub>l</sub>*.
> For each name *n* in the union of the *N<sub>l<sub>*s where *l* is in *L<sub>s<sub>*:
> * If *L<sub>s</sub>* contains a deferred import (then it is a singleton set) and *n* is `loadLibrary`, then nothing further happens (the imported `loadLibrary` is shadowed by the special method introduced above).
> * Let *D* = {*d<sub>1</sub>*, ..., *d<sub>k</sub>*} be the set of declarations with name *n* introduced by the imports of *L<sub>s</sub>*.  Declarations are identified by their name and the library they originated in, so if the same declaration is introduced more than once, through duplicate imports or re-exports, it only counts once in this set. This set cannot be empty.
> * If *D* has only one element, *s* contains that element as its declaration for *n*.
> * Otherwise, if all except one element, *d*, of *D* is a member declared in a platform library (a library with a URI starting with `dart:`), then *s* contains *d* as its declaration for *n*.
> * Otherwise *s* contains a NAME_CLASH marker for *n*

(or something)

**Comment by eernstg** [comment]
Done in #1161.
